server/game/gameSystems/DiscardEntireHandSystem.ts

```diff
@@ -0,0 +1,44 @@
+import { EventName, Location } from '../core/Constants';
+import { AbilityContext } from '../core/ability/AbilityContext';
+import * as Helpers from '../core/utils/Helpers.js';
+import type Player from '../core/Player';
+import { IPlayerTargetSystemProperties, PlayerTargetSystem } from '../core/gameSystem/PlayerTargetSystem';
+import { DiscardSpecificCardSystem } from './DiscardSpecificCardSystem';
+
+export interface IDiscardEntireHandSystemProperties extends IPlayerTargetSystemProperties {
+    target: Player | Player[];
+}
+
+/**
+ * A {@link GameSystem} which discards a player's entire hand
+ */
+export class DiscardEntireHandSystem<TContext extends AbilityContext = AbilityContext> extends PlayerTargetSystem<TContext, IDiscardEntireHandSystemProperties> {
+    public override name = 'discardEntireHand';
+    public override eventName = EventName.OnEntireHandDiscarded;
+    public override readonly costDescription: string = 'discard hand';
+    public override readonly effectDescription: string = 'discard hand';
+
+    public override eventHandler(): void {
+        // Do nothing
+    }
+
+    public override queueGenerateEventGameSteps(events: any[], context: TContext, additionalProperties: Record<string, any> = {}): void {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties);
+
+        for (const player of Helpers.asArray(properties.target) as Player[]) {
+            // Ensure that the player can be affected by this system
+            if (!this.canAffect(player, context, additionalProperties)) {
```

Comment:
The way the pipeline is written we can assume that `canAffect` has been used to filter targets by the time we get to `queueGenerateEventGameSteps` or `eventHandler`, so we can assume that every target is already valid and don't need to call `canAffect` here

Reply:
The implementation of `queueGenerateEventGameSteps` in `GameSystem` checks for `canAffect` so that's why I added it. Happy to remove it if you think it's not needed.

Reply:
Ohh good catch, hmm. Let's leave it then, there's probably a reason.

Reply:
Done!